weeks/week02/section04.md

```diff
@@ -1,32 +1,289 @@
+```
+
+
+### Shallow copy
+
+Use `copy()` to create a shallow copy of a set.
+
+```python
+test_cases = {'TC01', 'TC02', 'TC03'}
+shallow_copy_of_test_cases = test_cases.copy()
+test_cases.add('TC04')
+shallow_copy_of_test_cases.add('TC05')
+print(test_cases) # Output: {'TC04', 'TC02', 'TC03', 'TC01'}
+print(shallow_copy_of_test_cases) # Output: {'TC05', 'TC02', 'TC03', 'TC01'}
+```
+
+
+## 4.4 Deleting
+
+You can remove elements from a `set` using `remove()`, `discard()`, `pop()` and `clear()` methods. Be cautious with `remove()` as it raises an error if the element is not found, whereas `discard()` will not.
+
+
+### `remove()`:
+
+Removes the specified element, but raises an error if it’s not found.
+
+```python
+colors = {'yellow', 'pink', 'green', 'black', 'red', 'blue'}
+colors.remove('red') # `colors` now is {'yellow', 'pink', 'green', 'black', 'blue'}
+colors.remove('red')  # It raises an exception similar to below
+
+"""
+Traceback (most recent call last):
+  File "C:\python\py311\Lib\code.py", line 90, in runcode
+    exec(code, self.locals)
+  File "<input>", line 1, in <module>
+KeyError: 'red'
+"""
+```
+
+
+### `discard()`:
+
+Removes the specified element, but does not raise an error if the element is not found.
+
+```python
+colors = {'yellow', 'pink', 'green', 'black', 'red', 'blue'}
+colors.discard('black') # `colors` now is {'green', 'pink', 'red', 'blue', 'yellow'}
+colors.discard('black') # Does nothing, no error
+
 ```
 
+### `pop()`
+
+Removes and returns an arbitrary element from the set.
+
+```python
+colors = {'green', 'pink', 'red', 'blue', 'yellow'}
+value_returned = colors.pop() # Removes and returns an arbitrary element
+print(value_returned) # Output: 'yellow'
+```
+
+
+### `clear()`
+
+Removes all elements from the set.
+
+```python
+colors = {'green', 'pink', 'red', 'blue', 'yellow'}
+colors.clear() # Removes and returns an arbitrary element
+print(colors) # Output: set()
+```
+
+
+### Mathematical set operations:
+
+Python sets support mathematical operations like `union`, `intersection`, `difference`, and `symmetric difference`, which are very efficient and commonly used in problem-solving.
+
+
+#### Union:
+
+Combines all unique elements from two or more sets.
+
+```python
+set_a = {1, 2, 3, 4}
+set_b = {3, 4, 5, 6}
+set_a.union(set_b) # Output: {1, 2, 3, 4, 5, 6}
+```
+
+
+#### Intersection:
+
+Finds the common elements between sets.
+
+```python
+set_a = {1, 2, 3, 4}
+set_b = {3, 4, 5, 6}
+set_a.intersection(set_b) # Output: {3, 4}
+```
+
+
+#### Difference
+
+Returns elements that are in the first `set` but not in the second.
+
+```python
+set_a = {1, 2, 3, 4}
+set_b = {3, 4, 5, 6}
+set_a.difference(set_b) # Output: {1, 2}
+```
+
+#### Symmetric Difference
+
+Returns elements that are in either `set` but not in both.
+
+```python
+set_a = {1, 2, 3, 4}
+set_b = {3, 4, 5, 6}
+set_a.symmetric_difference(set_b) # Output: {1, 2, 5, 6}
+```
+
+
+#### Additional Set Methods
+
```

Comment:
If something is expected to be here? If not - let's remove #### Additional Set Methods

Reply:
I added a description. There are a few methods written right after this subsection that are important to understand and are not directly related to math operations.